conans/client/remote_manager.py

```diff
@@ -133,37 +133,49 @@ def get_package(self, pref, layout, remote, output, recorder):
 
         output.info("Retrieving package %s from remote '%s' " % (pref.id, remote.name))
         layout.package_remove(pref)  # Remove first the destination folder
-        pkg_folder = layout.package(pref)
         with layout.set_dirty_context_manager(pref):
-            self._get_package(pref, pkg_folder, remote, output, recorder)
+            self._get_package(layout, pref, remote, output, recorder)
 
         self._hook_manager.execute("post_download_package", conanfile_path=conanfile_path,
                                    reference=pref.ref, package_id=pref.id, remote=remote)
 
-    def _get_package(self, pref, dest_folder, remote, output, recorder):
+    def _get_package(self, layout, pref, remote, output, recorder):
         t1 = time.time()
         try:
             pref = self._resolve_latest_pref(pref, remote)
             snapshot = self._call_remote(remote, "get_package_snapshot", pref)
             if not is_package_snapshot_complete(snapshot):
                 raise PackageNotFoundException(pref)
-            zipped_files = self._call_remote(remote, "get_package", pref, dest_folder)
 
-            package_checksums = calc_files_checksum(zipped_files)
+            download_pkg_folder = layout.download_package(pref)
+            # Download files to the pkg_tgz folder, not to the final one
+            zipped_files = self._call_remote(remote, "get_package", pref, download_pkg_folder)
```

Comment:
Somewhere around these lines, we need to set/unset the dirty flag to the downloaded file, it can fail during the download. I think the `get_package` call is only assigning the dirty flag to the package folder.
I see we are no longer removing the folder in the `except BaseException`

Reply:
That is the thing, as long as the whole package_folder is protected with a "dirty" flag, everything inside is dirty. There was not dirty flag for this download before, I would say it is not necessary. Probably the other dirty flags over individual .tgz files could also be removed, but I didn't want to change more things in this PR.

The removal of the folder also doesn't make sense, and apparently no test is covering it. If we have the "dirty" mechanism in place to protect interrupted processes, then it shouldn't be necessary to try to remove the folder (in some cases, it could even be annoying, as it won't let debug)

Reply:
Haven't tried it, but this is my concern:
 * Download starts
   + `package_folder` marked as dirty
 * Download fails
   + `package_folder` mark is still there
   + We have a broken file in `<layout>/download_folder/pid/xxxxxx.tgz` (not dirty)
 * Upload the package: lines in the uploader are checking the dirty for the tgz file, not the `package_folder`, will it upload a broken file?
   ```python
        download_pkg_folder = layout.download_package(pref)
        package_tgz = os.path.join(download_pkg_folder, PACKAGE_TGZ_NAME)
        if is_dirty(package_tgz):
            self._output.warn("%s: Removing %s, marked as dirty" % (str(pref), PACKAGE_TGZ_NAME))
            os.remove(package_tgz)
            clean_dirty(package_tgz)
       ... 
   ```

Without running the test I'm not sure if this is something that can happen or not.

Reply:
The code is doing both:

```python
    def _compress_package_files(self, layout, pref, integrity_check):
        t1 = time.time()
        if layout.package_is_dirty(pref):
            raise ConanException("Package %s is corrupted, aborting upload.\n"
                                 "Remove it with 'conan remove %s -p=%s'"
                                 % (pref, pref.ref, pref.id))

        download_pkg_folder = layout.download_package(pref)
        package_tgz = os.path.join(download_pkg_folder, PACKAGE_TGZ_NAME)
        if is_dirty(package_tgz):
            self._output.warn("%s: Removing %s, marked as dirty" % (str(pref), PACKAGE_TGZ_NAME))
            os.remove(package_tgz)
            clean_dirty(package_tgz)
```

First, it will raise if the package is dirty (as it doesn't know how to recover from that). If only the .tgz is dirty, it can try to remove it and re-compress.

Reply:
The test in ``test_upload_dirty()`` covers the case where the package is marked as dirty.

Reply:
I see, we are covered with that `raise` when the package_folder is dirty. 👍